sigma_delta.h

```diff
@@ -0,0 +1,72 @@
+/* 
+ sigma_delta.h - esp8266 sigma-delta source
+
+ Copyright (c) 2014 Ivan Grokhotkov. All rights reserved.
+ This file is part of the esp8266 core for Arduino environment.
+ 
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2.1 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+ 
+ /*
+/******************************************************************************
+ * Info Sigma delta module
+
+This module controls the esp8266 internal sigma delta source
+Each pin can be connected to the sigma delta source
+The target duty and frequency can be modified via the register GPIO_SIGMA_DELTA
+
+THE TARGET FREQUENCY IS DEFINED AS:
+
+FREQ = 80,000,000/prescaler * target /256  HZ,     0<target<128
+FREQ = 80,000,000/prescaler * (256-target) /256  HZ,     128<target<256
+target: duty ,0-255
+prescaler: clk_div,0-255
+so the target and prescale will both affect the freq.
+
+Usage :
+1. sigma_delta_enable() : activate the sigma delta source with default prescalar (0) & target (0)
+2. sigma_delta_attachPin(pin), any pin 0..15, TBC if gpio16 supports sigma-delta source
+     This will set the pin to NORMAL output mode (pinMode(pin,OUTPUT))
+3. sigma_delta_setPrescaler(uint8_t) : reduce the output frequencies
+4. sigma_delta_setTarget(uint8_t) : set the output signal duty cycle, duty cycle = target/256
+
+5. sigma_delta_detachPin(pin), this will revert the pin to NORMAL output mode & GPIO source. 
+The sigma delta source remains on until :
+6. sigma_delta_disable()
+
+*******************************************************************************/
+
+#ifndef SIGMA_DELTA_H
+#define SIGMA_DELTA_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void sigma_delta_enable(void);
+void sigma_delta_disable(void);
+void sigma_delta_attachPin(uint8_t pin);
+void sigma_delta_detachPin(uint8_t pin);
+bool sigma_delta_isPinAttached(uint8_t pin);
+uint8_t sigma_delta_getTarget(void);
+void sigma_delta_setTarget(uint8_t target);
+uint8_t sigma_delta_getPrescaler(void);
+void sigma_delta_setPrescaler(uint8_t prescaler);
+
```

Comment:
This is good, but I think to be most useful to end users, who may not be as familiar w/ clock dividers and sigma-delta operation a simpler API might be helpful.

I'd recommend renaming "setTarget/GetTarget" to simply, "write" or "read" as effectively they're trying to write an analog value and "target," while the name used by the hardware designers in the HDK, doesn't really mean much if you don't understand the error-diffusion behind things.

I'd also suggest a simpler "setFrequency(long f) setter method that would do the math described in the header and figure out the closest prescaler given the F_CPU.  That'd make end-user code simpler to understand.

Reply:
Not sure if I fully understand the setFrequency idea. Every call to 'write' (setTarget) would need to adapt the prescaler as well to fit this chosen frequency? And remove the setPrescaler/getPrescaler functions altogether?
There was some sample code like this in an earlier post to this issue.  In many cases the effective output frequency is way off the chosen frequency f. Suppose you chose 1 MHz. With 50% duty cycle the closest you get is 312.5kHz. So you give the user the impression that the generator can be used as a fixed frequency source, but in fact it cannot.

Reply:
Good point, I withdraw this silly request. :)  I was thinking this would be useful for someone to, say, drive a 4-pin PWM fan where they'd need to set frequency = 25khz, duty = %age, but it seems it's not really precise enough for such a use, anyway.

For the examples, please do add the Public Domain attribution.  I think the team is trying to get these examples cleaned up and that's an easy thing to do.

Reply:
Apparently ESP32 has an implementation similar to what you suggested. I modified the code to be compatible. ESP32 has apparently 8 independent channels, esp8266 has only 1. Apart from that, the behavior is the same (eg. in terms of output frequencies.) So it makes sense to align the programming interfaces.
Arduino-esp32 also provides a working sigma delta example; there it's not necessary to explicitly include the sigma_delta.h file in the example code; it would be convenient if this was the same in arduino-esp8266, but my example currently doesn't work without the explicit include